File: clients/android/app/src/main/java/com/personalvpn/client/WireGuardConfigRenderer.kt
package com.personalvpn.client

import java.net.URI
import java.util.Base64

/** Converts the control-plane response to the official WireGuard config format. */
object WireGuardConfigRenderer {
    fun render(privateKey: String, configuration: TunnelConfiguration): String {
        require(isKey(privateKey) && isKey(configuration.serverPublicKey)) { "invalid WireGuard key" }
        require(configuration.addresses.isNotEmpty() && configuration.allowedIps.isNotEmpty()) { "configuration has no addresses/routes" }
        val endpoint = URI("udp://${configuration.endpoint}")
        require(endpoint.host != null && endpoint.port in 1..65535 && endpoint.userInfo == null && endpoint.path.isEmpty() && endpoint.query == null && endpoint.fragment == null) { "invalid WireGuard endpoint" }
        val builder = StringBuilder()
            .append("[Interface]\nPrivateKey = ").append(privateKey).append('\n')
            .append("Address = ").append(configuration.addresses.joinToString(", ")).append('\n')
            .append("DNS = ").append(configuration.dnsServers.joinToString(", ")).append('\n')
            .append("\n[Peer]\nPublicKey = ").append(configuration.serverPublicKey).append('\n')
            .append("Endpoint = ").append(configuration.endpoint).append('\n')
            .append("AllowedIPs = ").append(configuration.allowedIps.joinToString(", ")).append('\n')
            .append("PersistentKeepalive = ").append(configuration.persistentKeepaliveSeconds).append('\n')
        require(configuration.persistentKeepaliveSeconds in 0..120) { "invalid keepalive" }
        return builder.toString()
    }

    private fun isKey(value: String): Boolean = try {
        Base64.getDecoder().decode(value).size == 32
    } catch (_: IllegalArgumentException) { false }
}

File: clients/windows/WireGuardConfigRenderer.cs
using System.Text;

namespace PersonalVpn.Client;

public static class WireGuardConfigRenderer
{
    public static string Render(string privateKey, VpnConfiguration configuration)
    {
        ArgumentNullException.ThrowIfNull(configuration);
        Validate(configuration);
        if (!IsKey(privateKey))
            throw new ArgumentException("Invalid WireGuard key", nameof(configuration));

        var builder = new StringBuilder()
            .AppendLine("[Interface]")
            .AppendLine($"PrivateKey = {privateKey}")
            .AppendLine($"Address = {string.Join(", ", configuration.Addresses)}")
            .AppendLine($"DNS = {string.Join(", ", configuration.DnsServers)}")
            .AppendLine()
            .AppendLine("[Peer]")
            .AppendLine($"PublicKey = {configuration.ServerPublicKey}")
            .AppendLine($"Endpoint = {configuration.Endpoint}")
            .AppendLine($"AllowedIPs = {string.Join(", ", configuration.AllowedIps)}")
            .AppendLine($"PersistentKeepalive = {configuration.PersistentKeepaliveSeconds}");
        return builder.ToString();
    }

    public static void Validate(VpnConfiguration configuration)
    {
        ArgumentNullException.ThrowIfNull(configuration);
        if (!IsKey(configuration.ServerPublicKey))
            throw new ArgumentException("Invalid WireGuard key", nameof(configuration));
        if (configuration.Addresses is null || configuration.AllowedIps is null || configuration.Addresses.Count == 0 || configuration.AllowedIps.Count == 0)
            throw new ArgumentException("Configuration has no addresses or routes", nameof(configuration));
        if (configuration.PersistentKeepaliveSeconds is < 0 or > 120)
            throw new ArgumentException("Invalid keepalive", nameof(configuration));

        Uri endpoint;
        try { endpoint = new Uri($"udp://{configuration.Endpoint}"); }
        catch (UriFormatException exception) { throw new ArgumentException("Invalid UDP endpoint", nameof(configuration), exception); }
        if (string.IsNullOrEmpty(endpoint.DnsSafeHost) || endpoint.Port is < 1 or > 65535 ||
            !string.IsNullOrEmpty(endpoint.UserInfo) || endpoint.AbsolutePath != "/" ||
            !string.IsNullOrEmpty(endpoint.Query) || !string.IsNullOrEmpty(endpoint.Fragment))
            throw new ArgumentException("Invalid UDP endpoint", nameof(configuration));
    }

    private static bool IsKey(string value)
    {
        try { return Convert.FromBase64String(value).Length == 32; }
        catch (ArgumentException) { return false; }
    }
}
